src/FileHelper.php

```diff
@@ -142,4 +156,403 @@ public static function createDirectory($path, $mode = 0775): bool
+                    if (!$destinationExists) {
+                        // delay creation of destination directory until the first file is copied to avoid creating empty directories
+                        static::createDirectory($destination, $options['dirMode'] ?? 0775);
+                        $destinationExists = true;
+                    }
+                    copy($from, $to);
+                    if (isset($options['fileMode'])) {
+                        @chmod($to, $options['fileMode']);
+                    }
+                } elseif (!isset($options['recursive']) || $options['recursive']) {
+                    // recursive copy, defaults to true
+                    static::copyDirectory($from, $to, $options);
+                }
+                if (isset($options['afterCopy'])) {
+                    \call_user_func($options['afterCopy'], $from, $to);
+                }
+            }
+        }
+
+        closedir($handle);
+    }
+
+    /**
+     * Normalize options.
+     *
+     * @param array $options raw options.
+     *
+     * @return array normalized options.
+     */
+    protected static function normalizeOptions(array $options): array
+    {
+        if (!array_key_exists('caseSensitive', $options)) {
+            $options['caseSensitive'] = true;
+        }
+
+        if (isset($options['except'])) {
+            foreach ($options['except'] as $key => $value) {
+                if (\is_string($value)) {
+                    $options['except'][$key] = self::parseExcludePattern($value, $options['caseSensitive']);
+                }
+            }
+        }
+
+        if (isset($options['only'])) {
+            foreach ($options['only'] as $key => $value) {
+                if (\is_string($value)) {
+                    $options['only'][$key] = self::parseExcludePattern($value, $options['caseSensitive']);
+                }
+            }
+        }
+
+        return $options;
+    }
+
+    /**
+     * Checks if the given file path satisfies the filtering options.
+     *
+     * @param string $path the path of the file or directory to be checked.
+     * @param array $options the filtering options.
+     *
+     * @return bool whether the file or directory satisfies the filtering options.
+     */
+    public static function filterPath(string $path, array $options): bool
+    {
+        if (isset($options['filter'])) {
+            $result = \call_user_func($options['filter'], $path);
+            if (\is_bool($result)) {
+                return $result;
+            }
+        }
+
+        if (empty($options['except']) && empty($options['only'])) {
+            return true;
+        }
+
+        $path = str_replace('\\', '/', $path);
+
+        if (!empty($options['except'])) {
+            $except = self::lastExcludeMatchingFromList($options['basePath'], $path, $options['except']);
+            if ($except !== null) {
+                return $except['flags'] & self::PATTERN_NEGATIVE;
```

Comment:
That won't return boolean.

Reply:
Add example in tests.

Reply:
So since Yii2 and YiiC pray, I will take a look at all the possible options, anyway any welcome suggestions.

Reply:
Yii 2 was not paying attention much to strict types returning 1 or 0 sometimes. That was acceptable back then.